src/rust.rs

```diff
@@ -463,6 +464,185 @@ macro_rules! rooted {
+    }
+    pub unsafe fn remove_from_root_stack(&mut self) {
+        self._base._base.remove_from_root_stack();
+    }
+}
+
+use jsapi::_vftable_CustomAutoRooter as CustomAutoRooterVFTable;
+/// CustomAutoRooters use dynamic dispatch on the C++ side for custom tracing,
+/// so provide trace logic via vftable when creating object on Rust side.
+unsafe trait CustomAutoTraceable: Sized {
+    const vftable: CustomAutoRooterVFTable = CustomAutoRooterVFTable {
+        trace: Self::trace,
+    };
+
+    unsafe extern "C" fn trace(this: *mut ::std::os::raw::c_void, trc: *mut JSTracer) {
+        let this = this as *const Self;
+        let this = this.as_ref().unwrap();
+        Self::do_trace(this, trc);
+    }
+
+    /// Used by `CustomAutoTraceable` implementee to trace its contents.
+    /// Corresponds to virtual `trace` call in CustomAutoRooter superclasses (C++).
+    fn do_trace(&self, trc: *mut JSTracer);
+}
+
+unsafe impl<T: CustomTrace> CustomAutoTraceable for CustomAutoRooter<T> {
+    fn do_trace(&self, trc: *mut JSTracer) {
+        self.data.as_ref().unwrap().trace(trc);
+    }
+}
+
+impl<T: CustomTrace> CustomAutoRooter<T> {
+    pub fn new() -> Self {
+        CustomAutoRooter {
+            _base: jsapi::CustomAutoRooter {
+                _vftable: &<Self as CustomAutoTraceable>::vftable,
+                _base: AutoGCRooter::new_unrooted(AutoGCRooterTag::CUSTOM),
+            },
+            data: None,
+        }
+    }
+}
+
+/// Similarly to `Trace` trait, it's used to specify tracing of various types
+/// that are used in conjunction with `CustomAutoRooter`.
+pub unsafe trait CustomTrace {
+    fn trace(&self, trc: *mut JSTracer);
+}
+
+unsafe impl CustomTrace for Vec<*mut JSObject> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for obj in self {
+            let obj = obj as *const *mut _ as *mut *mut _;
+            unsafe { CallObjectRootTracer(trc, obj, c_str!("sequence<object>")); }
+        }
+    }
+}
+
+unsafe impl CustomTrace for  Vec<Value> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for val in self {
+            let val = val as *const _ as *mut _;
+            unsafe { CallValueRootTracer(trc, val, c_str!("sequence<any>")); }
+        }
+    }
+}
+
+unsafe impl<T: CustomTrace> CustomTrace for Vec<Option<T>> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for seq in self.iter().filter_map(|x| x.as_ref()) {
+            seq.trace(trc);
+        }
+    }
+}
+
+unsafe impl<T: CustomTrace> CustomTrace for Vec<T> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for seq in self {
+            seq.trace(trc);
+        }
+    }
+}
+
+// TODO: Implement other CustomTraces
+// See: https://dxr.mozilla.org/mozilla-central/rev/ee21e5f7f1c1726e0ed2697eb45df54cdceedd36/dom/bindings/BindingUtils.h#2289
+
+pub type SequenceRooter<T> = CustomAutoRooter<Vec<T>>;
+
+/// TODO: Document me (similar to Rooted)
+pub struct SequenceRooterGuard<'a, T: 'a + RootKind + GCMethods> {
```

Comment:
For some of the work that the rest of your group is doing, we will want to make this more general. I'm fine if you want to make this follow-up work.

Since there is nothing special about sequences here, let's make this more general: `RooterGuard<T>` where `T: StackRoot` (which can be a new type that provides add_to_root_stack and remove_from_stack, as well as a way to access the contained data).